lua/lazyvim/plugins/extras/lang/java.lua

```diff
@@ -1,125 +1,143 @@
-                })
-              end
-              local jdtls_base_config = {
-                on_attach = require("lazyvim.util").on_attach(function(client, buffer)
-                  if mason_registry.has_package("java-test") then
-                    -- custom keymaps for Java test runner (not yet compatible with neotest)
-                    vim.keymap.set("n", "<leader>tT", function() require("jdtls").pick_test({ bufnr = buffer }) end, { buffer = buffer, desc = "Run specific Test" })
-                    vim.keymap.set("n", "<leader>tt", function() require("jdtls").test_class({ bufnr = buffer }) end, { buffer = buffer, desc = "Run File" })
-                    vim.keymap.set("n", "<leader>tr", function() require("jdtls").test_nearest_method({ bufnr = buffer }) end, { buffer = buffer, desc = "Run nearest" })
-                  end
-                  if mason_registry.has_package("java-debug-adapter") then
-                    -- custom init for Java debugger
-                    require("jdtls").setup_dap({ hotcodereplace = "auto" })
-                    require("jdtls.dap").setup_dap_main_class_configs()
-                  end
-                  require("jdtls.setup").add_commands()
-                end),
-                cmd = cmd,
-                root_dir = root_dir,
-                init_options = {
-                  bundles = bundles,
-                }
-              }
-              local jdtls_opts = require("lazyvim.util").opts("nvim-jdtls")
-              require("jdtls").start_or_attach(vim.tbl_deep_extend("force", jdtls_opts or {}, jdtls_base_config))
-              require("which-key").register({ c = { x = { name = "Extract" } } }, { prefix = "<leader>" })
-            end,
-          })
-          return true -- avoid duplicate servers
+          return true
         end,
       },
     },
   },
+  {
+    "mfussenegger/nvim-jdtls",
+    dependencies = {
+      "neovim/nvim-lspconfig",
+      "mfussenegger/nvim-dap",
+      "hrsh7th/nvim-cmp",
+      "folke/which-key.nvim",
+    },
+    ft = { "java" },
+    config = function()
+      local project_name = vim.fn.fnamemodify(vim.fn.getcwd(), ":p:h:t")
+      local workspace_dir = vim.fn.stdpath("data") .. "/lsp/java/" .. project_name
+      local config = {
+        cmd = {
+          "java",
+          "-Declipse.application=org.eclipse.jdt.ls.core.id1",
+          "-Dosgi.bundles.defaultStartLevel=4",
+          "-Declipse.product=org.eclipse.jdt.ls.core.product",
+          "-Dlog.protocol=true",
+          "-Dlog.level=ALL",
+          "-Xmx1g",
+          "--add-modules=ALL-SYSTEM",
+          "--add-opens",
+          "java.base/java.util=ALL-UNNAMED",
+          "--add-opens",
+          "java.base/java.lang=ALL-UNNAMED",
+          "-jar",
+          vim.fn.expand("$MASON/share/jdtls/plugins/org.eclipse.equinox.launcher.jar"),
+          "-configuration",
+          vim.fn.expand("$MASON/share/jdtls/config"),
+          "-data",
+          workspace_dir,
+        },
```

Comment:
According to the jdtls doc(https://github.com/eclipse-jdtls/eclipse.jdt.ls#running-from-command-line-with-wrapper-script), the command "jdtls" would work if the user has Python 3.9 installed, can't guarantee that.

The more secure way I think is to use the "java -De...", because if someone enable the LazyVim Java support, the requirement of "java" existing in PATH is reasonable.

And more, exposing more detail can help user to understand what the config do, and help them to create their own config if they have more requirement(for example, "-xmx1g")

Reply:
Hmm, it's a shame that the script depends on Python (even if most people will have it installed)

I'm still uneasy about this change. The purpose of the script is to abstract some details so that a plugin such as this doesn't need to reiterate them. The args you used are slightly different from those produced by the Python script, for reasons that aren't explained. Additionally there is no way for the user to augment these arguments without completely overriding the config function, so the idea that the user can create their own config is somewhat lacking.

I don't mean to be completely discouraging. It is possible that this is still the proper way forward. But you've just made a lot of changes without explaining them in comments or genuinely making things easier for the user to tune, which leads to asking lots of questions in code review.